EDGE-2053 Documentation for MetalLB FRR-K8s tech preview (#1125)
suse-edge/suse-edge.github.io: asciidoc/guides/metallb-k3s_frr-k8s.adoc

[#guides-metallb-k3s-frr-k8s]
= MetalLB on K3s (using FRR-K8s Mode)
:revdate: 2026-05-04
:page-revdate: {revdate}
:experimental:

ifdef::env-github[]
:imagesdir: ../images/
:tip-caption: :bulb:
:note-caption: :information_source:
:important-caption: :heavy_exclamation_mark:
:caution-caption: :fire:
:warning-caption: :warning:
endif::[]

MetalLB is a load-balancer implementation for bare-metal Kubernetes clusters, using standard routing protocols.

In this guide, we demonstrate how to deploy MetalLB in layer 3 FRR-K8s BGP mode.

== MetalLB on K3s (using FRR-K8s)
[NOTE]
====
FRR-K8s is currently a Technology Preview feature. Community documentation of
FRR-K8s is https://github.com/metallb/frr-k8s/[here].
====

In this quick start, FRR-K8s mode is used.

== Prerequisites

All prerequisites for FRR-K8s are the same as for <<guides-metallb-k3s-l3>>
apart from the need for a free IP address. The example here does not include
setting up a service so there is no need for an IP address. What is still needed
is: 

* A K3s cluster where MetalLB is going to be deployed.

> * Router(s) on the network that support the BGP protocol.

Reviewer: ```suggestion
* All prerequisites for FRR-K8s are the same as for <<guides-metallb-k3s-l3>>
apart from the need for a free IP address. 
+
[NOTE]
====
The example here does not include setting up a service so there is no need for an IP address. 
====
* A K3s cluster where MetalLB is going to be deployed.
* Router(s) on the network that support the BGP protocol.
```

== Configuration to Accept Incoming Route

> Out of the box MetalLB BGP advertises a Service IP address to all the peers that

Reviewer 2: I think "BGP peers" here would make it more clear (only here as it is the first time "peer" is mentioned; for the rest of the doc I think it is ok to keep only  "peer")

Author: Agree!

are configured. These peers, which are usually routers, will receive a route for
each Service IP address with a 32 bit network mask. When using FRR-K8s with the

> FRRConfiguration CR it is also possible to receive routes from external routers.

Reviewer: ```suggestion
FRRConfiguration CR, it is also possible to receive routes from external routers.
```

These external routes will show up in each node's routing table. There are
multiple benefits from this, but the main one is that it eliminates the need to
manually update Linux routing tables on nodes when the external network changes.

> Specifically in cases where it is desirable to avoid sending traffic through the

Reviewer: ```suggestion
manually update Linux routing tables on nodes when the external network changes, specifically in cases where it is desirable to avoid sending traffic through the
```

default gateway.

== Deployment

We will be using the MetalLB Helm chart published as part of the SUSE Edge
solution. Note that the FRR-K8s is a sub-chart of MetalLB, and to enable it

> frrk8s.enabled has to be set to true. FRR-K8s also requires some elevated

Reviewer: ```suggestion
solution. Note that the FRR-K8s is a sub-chart of MetalLB, and to enable it, set `frrk8s.enabled` to `true`. FRR-K8s also requires some elevated
```

privileges on the namespace. 

[,bash,subs="attributes"]
----
kubectl create namespace metallb-system
kubectl label namespace metallb-system pod-security.kubernetes.io/enforce=privileged
kubectl label namespace metallb-system pod-security.kubernetes.io/audit=privileged
kubectl label namespace metallb-system pod-security.kubernetes.io/warn=privileged

> helm install metallb \

Reviewer 2: I would add the "--create-namespace" option to be consistent with the other "helm install" commands we have in the SUSE Edge/Telco Cloud doc

Author: The issue is that you also need the elevated privileges and you have to add them before you install FRR-K8s. Thoughts on how to do that directly with the HELM command?

  oci://registry.suse.com/edge/charts/metallb \
  --namespace metallb-system \
  --set frrk8s.enabled=true --set frrk8s.external=false

while ! kubectl wait --for condition=ready -n metallb-system $(kubectl get\
 pods -n metallb-system -l app.kubernetes.io/component=controller -o name)\
 --timeout=10s; do
 sleep 2
done
----

> Verify that you have 4 pods in the metallb-system namespace and that they are

Reviewer: ```suggestion
Verify that you have 4 pods in the `metallb-system` namespace and that they are
```

all running without a problem: 

[,bash,subs="attributes"]
----
k get pods -n metallb-system
NAME                                                      READY   STATUS    RESTARTS     AGE
metallb-controller-7fbfd8977d-m2q9t                       1/1     Running   0            46s
metallb-metallb-frr-k8s-9w7wl                             6/6     Running   0            46s
metallb-metallb-frr-k8s-webhook-server-5d9d67ffd6-8jqnc   1/1     Running   1 (6s ago)   46s
metallb-speaker-qx8bl                                     1/1     Running   0            46s
----

== Configuration

> . At this point, the installation is complete. Create an `FRRConfiguration` for

Reviewer: Did you mean deployment or installation?

Author: I think installation as you do a "helm install"

Reviewer: I wanted to understand this point: A this point, installation is complete. Which installation do you refer to?

Author: I referred to the helm install in 26.4. Maybe I should move the "At this point, the installation ..." to the end of 26.4 (Deployment). It is confusing with the terminology here. Usually when you add something (e.g. a daemonset, deployment, service, etc) you talk about deploying it. However, the package manager for Kubernetes is Helm and the Helm terminology is talking about Installing (e.g. helm install ...)

  FRR-K8s:

> [,bash]

Reviewer: ```suggestion
  FRR-K8s:
+
[,bash]
```

indentation issues

Author: There was a strange bullet one there and I removed that instead. Now the [,bash] thing is aligned with the text

Author: Hm, I see what you wanted. You wanted the two steps to be bullets

----
cat <<-EOF | kubectl apply -f -
apiVersion: frrk8s.metallb.io/v1beta1
kind: FRRConfiguration
metadata:
  name: frrdemo
  namespace: metallb-system
spec:
  bgp:
    routers:
    - asn: 64513
      neighbors:
        - address: 192.168.20.154
          asn: 64512
          port: 179
          toAdvertise:
            allowed:
              mode: all
          toReceive:
            allowed:
              mode: all
EOF
----

> The FRR router has ASN 64512 while our `BGPPeer` will have 64513. We can also

Reviewer 2: According to yaml above:
- is not the ASN of the (local) FRR router 64513 (instead of the 64512 this text says)?
- and so the ASN for the configured BGPpeer is 64512 (instead of 64513)

Author: The external Router FRR router has ASN 64512

> The externa BGP router has ASN 64512 while our `BGPPeer` will have 64513. We can also

Reviewer 2: typo: missing a final "l" in "externa BGP router" (to it becomes "external" ;-)

> The externa BGP router has ASN 64512 while our `BGPPeer` will have 64513. We can also

Reviewer: ```suggestion
+
The external BGP router has ASN 64512 while our `BGPPeer` will have 64513. We can also
```

> see that the FRR Router has an IP address that is 192.168.20.154.

Reviewer 2: According to yaml above:
- that 192.168.20.154 IPv4 addr ... is not the one assigned to the BGPpeer (but not the one for the (local) FRR Router)?

Author: The external FRR router has the IP address 192.168.20.154. This is straight from our demo

> see that the exteranl BGP Router has an IP address that is 192.168.20.154.

Reviewer 2: typo: "external" (it now says "exteranl")

Author: Fixed

> Verify that the FRRConfiguration has been deployed: 

Reviewer: ```suggestion
. Verify that the FRRConfiguration has been deployed: 
```

> [,bash]

Reviewer: ```suggestion
+
[,bash]
```

----
k get FRRConfiguration -A
NAMESPACE        NAME      AGE
metallb-system   frrdemo   4s

> ----
> 
> 
> [NOTE]

Reviewer: ```suggestion
----
+
[NOTE]
```

====

> The toReceive settings above will make your cluster accept all incoming routes. This

Reviewer: ```suggestion
The `toReceive` settings above will make your cluster accept all incoming routes. This
```

might not be advisable in a production environment as, depending on your
external routers, it can cause your nodes' routing tables to fill up. See
documentation on https://github.com/metallb/frr-k8s/[how to filter toReceive
further.]
====

> With FRR-K8s these are all the settings needed. Any route that your external FRR

Reviewer 2: As we agreed: to replace "FRR" by "BGP" so it says "Any route that the external BGP router ..." ;-)

Author: Fixed

router is receiving will be shared with your cluster and all the nodes will have

> their routing tables updated.  +

Reviewer: ```suggestion
their routing tables updated. 
```

The setup can be tested with what is described in <<guides-metallb-k3s-l3>>. In
order to combine these setups there are some requirements: 

* The FRR-K8s configuration is done on a separate cluster. This is required to
  avoid cluster internal routing.
* Changes to ASN IDs and to the Router IP address are required to match both
  setups.

With both these setups in place the following can be observed:

> * Once the Deployment and Service have been applied on the cluster described in
>   <<guides-metallb-k3s-l3>> the route to the service will be visible on the
>   external FRR Router.

Reviewer: ```suggestion
* Once the Deployment and Service have been applied on the cluster described in
  <<guides-metallb-k3s-l3>>, the route to the service will be visible on the
  external FRR Router.
```

* The route will be added to all nodes on the FRR-K8s cluster.

[NOTE]
====

> In starndard configurations of FRR Routing routes are shared with the next-hop

Reviewer: ```suggestion
In standard configurations of FRR Routing routes are shared with the next-hop
```

set to the IP Address of the FRR Router. To achieve a next-hop without the FRR

> Router IP Address one can add a line "neighbor BGPPG next-hop-unchanged" and a

Reviewer: ```suggestion
set to the IP Address of the FRR Router. To achieve a next-hop without the FRR
Router IP Address, one can add a line "neighbor BGPPG next-hop-unchanged" and a
```

> line "neighbor BGPPG as-override" to the /etc/frr/frr.conf file on the FRR

Reviewer: ```suggestion
line "neighbor BGPPG as-override" to the `/etc/frr/frr.conf` file on the FRR
```

> Router. With this in place the nodes on the FRR-K8s cluster will get a direct

Reviewer: ```suggestion
Router. With this in place, the nodes on the FRR-K8s cluster will get a direct
```

[#guides-metallb-k3s-frr-k8s]
= MetalLB on K3s (using FRR-K8s Mode)
:revdate: 2026-05-04
:page-revdate: {revdate}
:experimental:

ifdef::env-github[]
:imagesdir: ../images/
:tip-caption: :bulb:
:note-caption: :information_source:
:important-caption: :heavy_exclamation_mark:
:caution-caption: :fire:
:warning-caption: :warning:
endif::[]

MetalLB is a load-balancer implementation for bare-metal Kubernetes clusters, using standard routing protocols.

In this guide, we demonstrate how to deploy MetalLB in layer 3 FRR-K8s BGP mode.

== MetalLB on K3s (using FRR-K8s)
[NOTE]
====
FRR-K8s is currently a Technology Preview feature. Community documentation of
FRR-K8s is https://github.com/metallb/frr-k8s/[here].
====

In this quick start, FRR-K8s mode is used.

== Prerequisites

* All prerequisites for FRR-K8s are the same as for <<guides-metallb-k3s-l3>>

> apart from the need for a free IP address.
> 
> [NOTE]

Reviewer: ```suggestion
apart from the need for a free IP address.
+
[NOTE]
```
Add the "+" sign to align indentation
